be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/MessageDecoder.scala

```diff
@@ -2,15 +2,31 @@ package ch.epfl.pop.pubsub.graph
 
 import akka.NotUsed
 import akka.stream.scaladsl.Flow
-import ch.epfl.pop.model.network.{JsonRpcRequest, MethodType}
+import ch.epfl.pop.model.network.{JsonRpcRequest, JsonRpcResponse}
+import ch.epfl.pop.jsonNew.HighLevelProtocol._
+import spray.json._
+
+import scala.util.{Success, Try}
 
 object MessageDecoder {
 
-  // Message or TextMessage
   val jsonRpcParser: Flow[Either[JsonString, PipelineError], GraphMessage, NotUsed] = Flow[Either[JsonString, PipelineError]].map {
-    case Left(jsonString) =>
-      Left(JsonRpcRequest("2.0", MethodType.INVALID, ???, Some(0)))
-    // case _ => _ // TODO check if type ok. Else prob Right(pipelineError)
+    case Left(jsonString) => Try(jsonString.parseJson.asJsObject) match {
+      case Success(obj) =>
+        val fields: Set[String] = obj.fields.keySet
+
+        if (fields.contains("method")) { // FIXME check that the error (if any) is correctly propagated
+          Left(obj.convertTo[JsonRpcRequest])
+        } else {
+          Left(obj.convertTo[JsonRpcResponse])
```

Comment:
I'm a bit confused as to the meaning of Left() and Right(), are we indicating that Left() is the left option in the `Either[JsonString, PipelineError]` ? Or should this be Right? 
Apologies for my Scala/Akka ignornace :)

Reply:
Everything (except the parser input) that goes through the graph is a `GraphMessage` that I defined as `Either[JsonRpcMessage, PipelineError]` (which has type parameters inverted as Gaurav mentionned... woops! ^^'). So everything that goes out of the parser is an instance of this weird "Either" object. If I want to output (success) a `JsonRpcMessage`, then I would use `Left(message)` to create this Either instance. If a problem occured, then I would use `Right(PipelineError(...))`

So in short, yes, Left creates an instance of `Either[V, W]` which holds an object of type `V`

PS: warning `Either[JsonString, PipelineError]` is the input, not the output!